src/hardware/pslab/api.c

```diff
@@ -46,154 +87,354 @@ static GSList *scan(struct sr_dev_driver *di, GSList *options)
 	GSList *device_paths = g_slist_concat(device_paths_v5, device_paths_v6);
 
 	devices = NULL;
-	struct sr_serial_dev_inst *serial;
-	struct sr_dev_inst *sdi;
-
-	const char *path = NULL;
-	const char *serialcomm = "1000000/8n1";
 
-	for (l = options; l; l = l->next)
-	{
+	for (l = options; l; l = l->next) {
 		src = l->data;
 		switch (src->key) {
-			case SR_CONF_CONN:
-				path = g_variant_get_string(src->data, NULL);
-				break;
-			case SR_CONF_SERIALCOMM:
-				serialcomm = g_variant_get_string(src->data, NULL);
-				break;
+		case SR_CONF_CONN:
+			path = g_variant_get_string(src->data, NULL);
+			break;
+		case SR_CONF_SERIALCOMM:
+			serialcomm = g_variant_get_string(src->data, NULL);
+			break;
 		}
 	}
 
-	for (l = device_paths; l; l = l->next)
-	{
-		char *device_path = l->data;
-		if (path && path != device_path) {
+	for (l = device_paths; l; l = l->next) {
+		device_path = l->data;
+		if (path && path != device_path)
 			continue;
-		}
+
 		serial = sr_serial_dev_inst_new(device_path, serialcomm);
-		if (serial_open(serial, SERIAL_RDWR) != SR_OK) {
+
+		if (serial_open(serial, SERIAL_RDWR) != SR_OK)
 			continue;
-		}
 
-		char* version = pslab_get_version(serial, COMMON, VERSION_COMMAND);
-		gboolean isPSLabDevice = g_str_has_prefix(version, "PSLab") || g_str_has_prefix(version, "CSpark");
-		if(!isPSLabDevice) {
+		version = pslab_get_version(serial);
+		gboolean isPSLabDevice = g_str_has_prefix(version, "PSLab") ||
+			g_str_has_prefix(version, "CSpark");
+		if (!isPSLabDevice) {
 			g_free(version);
 			serial_close(serial);
 			continue;
 		}
 		sr_info("PSLab device found: %s on port: %s", version, device_path);
 
 		sdi = g_new0(struct sr_dev_inst, 1);
+		devc = g_new0(struct dev_context, 1);
 		sdi->status = SR_ST_INACTIVE;
 		sdi->inst_type = SR_INST_SERIAL;
 		sdi->vendor = g_strdup("FOSSASIA");
 		sdi->connection_id = device_path;
 		sdi->conn = serial;
 		sdi->version = version;
 
-		struct sr_channel_group *cg = g_new0(struct sr_channel_group, 1);
-		cg->name = g_strdup("Analog");
-		cg->channels = g_slist_alloc();
-		for (int i = 0; i < NUM_ANALOG_CHANNELS; i++)
-		{
-			struct sr_channel *ch = sr_channel_new(sdi, i, SR_CHANNEL_ANALOG, TRUE, analog_channels[i].name);
+		for (i = 0; i < NUM_ANALOG_CHANNELS; i++) {
+			struct sr_channel *ch = sr_channel_new(sdi, analog_channels[i].index,
+				SR_CHANNEL_ANALOG, TRUE, analog_channels[i].name);
 			struct channel_priv *cp = g_new0(struct channel_priv, 1);
+			struct sr_channel_group *cg = g_new0(struct sr_channel_group, 1);
+			struct channel_group_priv *cgp =  g_new0(struct channel_group_priv, 1);
```

Comment:
Ideally, we should define the struct in the start of the function. I missed this. Nothing crtical. I will make a commit and rectify it before we merge. Not changing it right now as it might break the flow of somebody already reviewing the code

Reply:
Changing this too with the commit.